src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,940 @@
+  if GAP.Globals.IsDoneIterator(state)
+    return nothing
+  end
+  i = GAP.Globals.NextIterator(state)
+  return MatrixGroupElem(G, G.mat_iso(i),i), state
+end
+
+########################################################################
+#
+# Membership
+#
+########################################################################
+
+
+function ==(G::MatrixGroup,H::MatrixGroup)
+   G.deg==H.deg || return false
+   G.ring==H.ring || return false
+   if isdefined(G, :descr) && isdefined(H, :descr)
+      return G.descr == H.descr
+   end
+   if isdefined(G, :gens) && isdefined(H, :gens)
+      if G.gens==H.gens return true  end       #TODO this works only if the == between elements does not keep track of the parent
+   end
+   return G.X==H.X
+end
+
+
+# this saves the value of x.X
+# x_gap = x.X if this is already known, x_gap = nothing otherwise
+function lies_in(x::MatElem, G::MatrixGroup, x_gap)
+   if base_ring(x)!=G.ring || nrows(x)!=G.deg return false, x_gap end
+   if isone(x) return true, x_gap end
+   if isdefined(G,:gens)
+      for g in G.gens
+         if x==g.elm
+            return true, x_gap
+         end
+      end
+   end
+   if isdefined(G,:descr) && G.descr==:GL
+      if det(x)!=0
+         return true, x_gap
+      else
+         return false, x_gap
+      end
+   elseif isdefined(G,:descr) && G.descr==:SL
+      if det(x)==1
+         return true, x_gap
+      else
+         return false, x_gap
+      end
+   else
+      if x_gap==nothing x_gap = G.mat_iso(x) end
+     # x_gap !=nothing || x_gap = G.mat_iso(x)
+      return GAP.Globals.in(x_gap,G.X), x_gap
+   end
+end
+
+Base.in(x::MatElem, G::MatrixGroup) = lies_in(x,G,nothing)[1]
+
+function Base.in(x::MatrixGroupElem, G::MatrixGroup)
+   if isdefined(x,:X) return lies_in(x.elm,G,x.X)[1]
+   else return lies_in(x.elm,G,nothing)[1]
+   end
+end
+
+# embedding an element of type MatElem into a group G
+function (G::MatrixGroup)(x::MatElem)
+   vero, x_gap = lies_in(x,G,nothing)
+   vero || throw(ArgumentError("Element not in the group"))
+   if x_gap==nothing return MatrixGroupElem(G,x)
+   else return MatrixGroupElem(G,x,x_gap)
+   end
+end
+
+# embedding an element of type MatrixGroupElem into a group G
+function (G::MatrixGroup)(x::MatrixGroupElem)
+   if isdefined(x,:X)
+      if isdefined(x,:elm)
+         vero = lies_in(x.elm,G,x.X)[1]
+         vero || throw(ArgumentError("Element not in the group"))
+         return MatrixGroupElem(G,x.elm,x.X)
+      else
+         GAP.Globals.in(x.X, G.X) || throw(ArgumentError("Element not in the group"))
+         return MatrixGroupElem(G,x.X)
+      end
+   else
+      vero, x_gap = lies_in(x.elm,G,nothing)
+      vero || throw(ArgumentError("Element not in the group"))
+      if x_gap==nothing return MatrixGroupElem(G,x.elm)
+      else return MatrixGroupElem(G,x.elm,x_gap)
+      end
+   end
+end
+
+# embedding a nxn array into a group G
+function (G::MatrixGroup)(L::AbstractVector)
```

Comment:
Yes, it's a matrix. It's just a help, so if one wants to write an explicit matrix in the group `G`, he can just write `G([a,b,c,d])` instead of `G(matrix(F,2,2,[a,b,c,d]))`.

Reply:
> he can just write more concisely `G([a,b,c,d])` instead of `G(matrix(F,2,2,[a,b,c,d]))`.

Is it meant mostly for literal arrays? Then maybe it would be useful (instead or in addition to) to allow passing an `AbstractMatrix`, which can be entered as succintly as a vector (`G[a b; c d]`). (I personally prefer the `AbstractMatrix` version, which is more explicit, in particular regarding row vs. column major order). If you want to keep both, you could relax `L::AbstractVector` to `L::AbstractVecOrMat`.